### ISSUE TITLE
custom cache key

### ISSUE THREAD

**Issue by kevinranks** [dream-sports-labs/react-native-fast-image#158]
It would be awesome to have the ability to provide a custom `cacheKey` that would fall back to use the URI if none provided.

Our use-case is that some of the images are already available locally, prior to being uploaded to a remote distribution endpoint. 

With a custom `cacheKey` we could take advantage of the local images by setting something like like `cacheKey: imageID-${updatedAt}`  with the local image path. Later if it uses the remote URL it would already have the image cached based on the custom `cacheKey` and not need to re-cache the same image.

**Comment by princektripathi** [comment]
> It would be awesome to have the ability to provide a custom `cacheKey` that would fall back to use the URI if none provided.
> 
> Our use-case is that some of the images are already available locally, prior to being uploaded to a remote distribution endpoint.
> 
> With a custom `cacheKey` we could take advantage of the local images by setting something like like `cacheKey: ${image.id}-${updatedAt}` with the local image path. Later if it uses the remote URL it would already have the image cached based on the custom `cacheKey` and not need to re-cache the same image.

@kevinranks, thanks for your suggestion. 

I'm summarising my understanding so that we can brainstorm it internally and come up with a solution.

You expect an additional prop `cacheKey` to avoid caching redundant images. Currently, path based caching can lead to caching the same image twice specifically in scenarios where the image was loaded from local and later retrieved from network.

Have I understood it correctly?

**Comment by kevinranks** [comment]
> > It would be awesome to have the ability to provide a custom `cacheKey` that would fall back to use the URI if none provided.
> > Our use-case is that some of the images are already available locally, prior to being uploaded to a remote distribution endpoint.
> > With a custom `cacheKey` we could take advantage of the local images by setting something like like `cacheKey: ${image.id}-${updatedAt}` with the local image path. Later if it uses the remote URL it would already have the image cached based on the custom `cacheKey` and not need to re-cache the same image.
> 
> @kevinranks, thanks for your suggestion.
> 
> I'm summarising my understanding so that we can brainstorm it internally and come up with a solution.
> 
> You expect an additional prop `cacheKey` to avoid caching redundant images. Currently, path based caching can lead to caching the same image twice specifically in scenarios where the image was loaded from local and later retrieved from network.
> 
> Have I understood it correctly?

Yes exactly. In an app that users upload many images to a remote server, it makes sense to be able to utilize the local copy of that image prior to uploading it remotely. It's my understanding that right now the internal cache key is based off of the URL. By offering the ability to override that it gives developers a lot more control over the cache. For example, being able to reference a local path for an image and then later use the remote URL, and as long as the cacheKey is the same for both you can show the local (temp) image while the image is being uploaded remotely, all without any additional components or special logic.

**Comment by princektripathi** [comment]
@kevinranks, it definitely makes sense. We will brainstorm internally and come up with a solution. Also, if you're interested in contributing to this repo, would suggest you to join our discord channel - [https://discord.gg/BNHhcpYD](url)

**Comment by Pauligrinder** [comment]
Any progress on this? I'd be glad to help out if needed. For us, this would be used such that the cache-key would be set to the image url without parameters -> our image hosting has very short-lived tokens, and therefore our api returns a new generated token each time (included in the image url). And then of course the image component thinks it's a new image every time.

react-native-turbo-image has this functionality, where you can put "cache-key": string into the source={}. Unfortunately that library has major performance issues in my app, so I came back to fast-image because of that.

There was also a fork of Fast-Image that had a "ignore url params in cache key" but that one never got maintained.
